interactors/nonceHandlerV3/addressNonceHandler.go

```diff
@@ -92,26 +108,37 @@ func (anh *addressNonceHandler) Close() {
 	anh.cancelFunc()
 }
 
-func (anh *addressNonceHandler) fetchGasPriceIfRequired(ctx context.Context) {
+func (anh *addressNonceHandler) applyGasPriceIfRequired(ctx context.Context, tx *transaction.FrontendTransaction) {
+	anh.mut.Lock()
+	defer anh.mut.Unlock()
+
 	if anh.gasPrice == 0 {
 		networkConfig, err := anh.proxy.GetNetworkConfig(ctx)
 
-		anh.mut.Lock()
-		defer anh.mut.Unlock()
 		if err != nil {
 			log.Error("%w: while fetching network config", err)
 			anh.gasPrice = 0
 			return
 		}
 		anh.gasPrice = networkConfig.MinGasPrice
 	}
+	tx.GasPrice = core.MaxUint64(anh.gasPrice, tx.GasPrice)
 }
 
-func (anh *addressNonceHandler) fetchNonce(ctx context.Context) (uint64, error) {
-	account, err := anh.proxy.GetAccount(ctx, anh.address)
-	if err != nil {
-		return 0, err
-	}
+func (anh *addressNonceHandler) fetchNonce(ctx context.Context) (int64, error) {
+	// if it is the first time applying nonces to this address, or if the cache was invalidated it will try to fetch
+	// the nonce from the chain.
+	anh.mut.Lock()
```

Comment:
also here. The mutex is kept longer than necessary.

Reply:
changed.

Reply:
yup, true. I guess this should remain like this as it only keeps the lock only if the fetch is needed. Which should block any other access as to have consistent values.